hybris-fb.c

```diff
@@ -166,27 +136,84 @@ hybris_device_fb_init(void)
 
   done = true;
 
-  hybris_plugin_fb_open_device(&hybris_device_fb_handle);
-
-  if( !hybris_device_fb_handle ) {
-    mce_log(LL_ERR, "failed to open framebuffer device");
+  if( !hybris_plugin_fb_load() ) {
     goto cleanup;
   }
 
-  mce_log(LL_DEBUG, "hybris_device_fb_handle = %p",
-          hybris_device_fb_handle);
+  /* Open frame buffer device */
+  if( hybris_plugin_fb_handle ) {
+    hybris_plugin_fb_handle->methods->open(hybris_plugin_fb_handle,
+                                           GRALLOC_HARDWARE_FB0,
+                                           (hw_device_t**)&hybris_device_fb_handle);
+    if( !hybris_device_fb_handle ) {
+      mce_log(LL_WARN, "failed to open frame buffer device");
+    }
+  }
+
+  /* Open hw composer device */
+  if( hybris_plugin_hwc_handle ) {
+    hybris_plugin_hwc_handle->methods->open(hybris_plugin_hwc_handle,
+                                            HWC_HARDWARE_COMPOSER,
+                                            (hw_device_t**)&hybris_device_hwc_handle);
+    if( !hybris_device_hwc_handle ) {
+      mce_log(LL_WARN, "failed to open hw composer device");
+    }
+  }
+
+  /* What we'd like to use is:
+   *
+   * 1. hwc_dev->setPowerMode()
+   * 2. hwc_dev->blank()
+   * 2. fb_dev->enableScreen()
+   *
+   * While all are optional, having none available is unexpected.
+   */
+
+  if( hybris_device_hwc_handle ) {
+#ifdef HWC_DEVICE_API_VERSION_1_4
+    if( hybris_device_hwc_handle->common.version >= HWC_DEVICE_API_VERSION_1_4 &&
+        hybris_device_hwc_handle->setPowerMode ) {
+      mce_log(LL_DEBUG, "using hw composer setPowerMode() method");
+      ack = true;
+      goto cleanup;
+    }
+#endif
+    if( hybris_device_hwc_handle->blank ) {
+      mce_log(LL_DEBUG, "using hw composer blank() method");
+      ack = true;
+      goto cleanup;
+    }
+  }
+
+  if( hybris_device_fb_handle ) {
+    if( hybris_device_fb_handle->enableScreen ) {
+      mce_log(LL_DEBUG, "using framebuffer enableScreen() method");
+      ack = true;
+      goto cleanup;
+    }
+  }
+
+  mce_log(LL_WARN, "no known display power control interfces");
 
 cleanup:
 
-  return hybris_device_fb_handle != 0;
+  return ack;
 }
 
 /** Release libhybris frame buffer device object
  */
 void
 hybris_device_fb_quit(void)
 {
-  hybris_plugin_fb_close_device(&hybris_device_fb_handle);
+  if( hybris_device_hwc_handle ) {
+    hybris_device_hwc_handle->common.close(&hybris_device_hwc_handle->common);
+    hybris_device_hwc_handle = 0;
+  }
+
+  if( hybris_device_fb_handle ) {
+    hybris_device_fb_handle->common.close(&hybris_device_fb_handle->common);
+    hybris_device_fb_handle = 0;
```

Comment:
android_dlclose(hybris_device_fb_handle->common.dso); additionally?

also judging from comments in android files we might need a no_unload quirk here someday.

(same comment applies few lines above)